Nextion update (#934)
OttoWinter/esphomedocs: components/display/nextion.rst

.. figure:: images/nextion-full.jpg
    :align: center

>     :width: 75.0%    

Reviewer: ```suggestion
    :width: 75.0%
```

  See :ref:`display-nextion_lambda` for more information. This is typically empty. The individual components for the Nextion will handle almost all features needed for updating
- **update_interval** (*Optional*, :ref:`config-time`): The interval to call the lambda to update the display.
  Defaults to ``5s``.
- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
- **tft_url** (*Optional*, string): The URL to download the TFT file from for updates. See :ref:`Nextion Upload <nextion_upload_tft>`.
- **on_sleep** (*Optional*, :ref:`Action <config-action>`): An automation to perform when the Nextion goes to sleep.
- **on_wake** (*Optional*, :ref:`Action <config-action>`): An automation to perform when the Nextion wakes up.
- **touch_sleep_timeout** (*Optional*, int): Sets internal No-touch-then-sleep timer in seconds.
- **wake_up_page** (*Optional*, int): Sets the page to display after waking up
- **auto_wake_on_touch** (*Optional*, boolean): Sets if Nextion should auto-wake from sleep when touch press occurs.
  

.. _display-nextion_lambda:

> Rendering Lambda

Reviewer: ```suggestion

Rendering Lambda
```

  Use this if you want to use multiple UART buses at once.
- **brightness** (*Optional*, percentage): Set display brightness in %. Defaults to ``100%``
- **lambda** (*Optional*, :ref:`lambda <config-lambda>`): The lambda to use for rendering the content on the nextion display.
  See :ref:`display-nextion_lambda` for more information. This is typically empty. The individual components for the Nextion will handle almost all features needed for updating
- **update_interval** (*Optional*, :ref:`config-time`): The interval to call the lambda to update the display.
  Defaults to ``5s``.
- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
- **tft_url** (*Optional*, string): The URL to download the TFT file from for updates. See :ref:`Nextion Upload <nextion_upload_tft>`.
- **on_sleep** (*Optional*, :ref:`Action <config-action>`): An automation to perform
  when the Nextion goes to sleep. See :ref:`Nextion On_Sleep/On_Wake <nextion_on_sleep_on_wake>`.
- **on_wake** (*Optional*, :ref:`Action <config-action>`): An automation to perform

>   when the Nextion wakes up. See :ref:`Nextion On_Sleep/On_Wake <nextion_on_sleep_on_wake>`.

Reviewer 2: ```suggestion
- **on_sleep** (*Optional*, :ref:`Action <config-action>`): An automation to perform when the Nextion goes to sleep.
- **on_wake** (*Optional*, :ref:`Action <config-action>`): An automation to perform when the Nextion wakes up.
```

- **touch_sleep_timeout** (*Optional*, int): Sets internal No-touch-then-sleep timer in seconds.
- **wake_up_page** (*Optional*, int): Sets the page to display after waking up

> - **touch_sleep_timeout** (*Optional*, int): Sets internal No-touch-then-sleep timer in seconds.

Reviewer 2: ```suggestion
```

- **auto_wake_on_touch** (*Optional*, boolean): Sets if Nextion should auto-wake from sleep when touch press occurs.
  

.. _display-nextion_lambda:

>  

Reviewer 2: ```suggestion

```

      on_wake:
        then:
          lambda: 'ESP_LOGD("display","Display woke up");'

.. _nextion_update_all_components:

- ``update_all_components()``: All the components will publish their states.

  .. code-block:: c++

      id(nextion1).update_all_components();

.. _update_components_by_prefix:

- ``update_components_by_prefix(std::string page)``: This will send the current state of any **component_name** matching the prefix. Some settings like background color need to be resent on page change. This is a good hook for that.

  .. code-block:: c++

      id(nextion1).update_components_by_prefix("page0.");

.. _set_nextion_sensor_state:

- ``set_nextion_sensor_state(NextionQueueType queue_type, std::string name, float state);`` : Sets the sensor state. See :ref:`Queue Types <nextion_queue_types>`
- ``set_nextion_sensor_state(int queue_type, std::string name, float state);`` : Sets the sensor state. See :ref:`Queue Types <nextion_queue_types>`

- ``set_nextion_text_state(std::string name, std::string state);`` : Sets the text sensor state

  .. note::

  Below is a method for HASS to send updates to the Nextion by code.

>   .. code-block:: yaml
> 
>   # Enable Home Assistant API
>   api:
>     services:
>       - service: set_nextion_sensor
>         variables:
>           nextion_type: int
>           name: string
>           state: float
>         then:
>           - lambda: |-
>               id(nextion1).set_nextion_sensor_state(nextion_type,name,state);
>       - service: set_nextion_text
>         variables:
>           name: string
>           state: string
>         then:
>           - lambda: |-
>               id(nextion1).set_nextion_text_state(name,state);

Reviewer: ```suggestion

  .. code-block:: yaml

    # Enable Home Assistant API
    api:
      services:
        - service: set_nextion_sensor
          variables:
            nextion_type: int
            name: string
            state: float
          then:
            - lambda: |-
                id(nextion1).set_nextion_sensor_state(nextion_type,name,state);
        - service: set_nextion_text
          variables:
            name: string
            state: string
          then:
            - lambda: |-
                id(nextion1).set_nextion_text_state(name,state);
```

          // set the text of a component with formatting
          it.set_component_text_printf("textview", "The uptime is: %.1f", id(uptime_sensor).state);

.. note::

>     Although you can use the rendering lambda most, if not all, updates to the Nextion can be handled by the individual Nextiom components. **See Below**

Reviewer: ```suggestion
    Although you can use the rendering lambda most, if not all, updates to the Nextion can be handled by the individual Nextion components. **See Below**
```

Please see :ref:`display-printf` for a quick introduction into the ``printf`` formatting rules and
:ref:`display-strftime` for an introduction into the ``strftime`` time formatting.

Lambda Calls
************

Several methods are available for use within :ref:`lambdas <config-lambda>` ; these permit advanced functionality beyond simple
display updates. See the full :apiref:`nextion/nextion.h` for more info. 

.. _nextion_upload_tft:

- ``upload_tft``: Start the upload process. See :ref:`nextion_upload_tft_file`

> The developer tools in Home Assitant can be used to trigger the update. The below code block is an example on how to set this up.

Reviewer 2: ```suggestion
The developer tools in Home Assistant can be used to trigger the update. The below code block is an example on how to set this up.
```

  .. code-block:: yaml

      api:
        services:
          - service: update_nextion
            then:
              - lambda: 'id(nextion1)->upload_tft();' 

.. _nextion_on_sleep_on_wake:

- ``on_sleep``/``on_wake``: Actions to perform when the Nextion goes in and out of sleep.

  .. code-block:: yaml

      // Within lambda
      on_sleep:
        then:
          lambda: 'ESP_LOGD("display","Display went to sleep");'
      on_wake:
        then:
          lambda: 'ESP_LOGD("display","Display woke up");'

> 

Reviewer 2: ```suggestion
```

.. note::

    Below is a method for HASS to send updates to the Nextion by code.

    .. code-block:: yaml

        # Enable Home Assistant API
        api:
          services:
            - service: set_nextion_sensor
              variables:
                nextion_type: int
                name: string
                state: float
              then:
                - lambda: |-
                    id(nextion1).set_nextion_sensor_state(nextion_type,name,state);
            - service: set_nextion_text
              variables:
                name: string
                state: string
              then:
                - lambda: |-
                    id(nextion1).set_nextion_text_state(name,state);

.. _nextion_queue_types:

 Queue Types: 
  - SENSOR            0
  - BINARY_SENSOR     1
  - SWITCH            2
  - TEXT_SENSOR       3
  - WAVEFORM_SENSOR   4
  - NO_RESULT         5

.. _nextion_upload_tft_file:

Uploading A TFT File
--------------------
This will download the file from the tft_url and will transfer it over the UART to the Nextion.
Once completed both the MCU and Nextion will reboot. During this process esphome will be 

> unresponsive and no logging will take place. This uses the same protocol as the Nextion editor and only updates the changes of the TFT file. If HTTPS/SSL is enabled it will be about 1kB/sec.

Reviewer 2: ```suggestion
Once completed both the ESP and Nextion will reboot. During the upload process esphome will be 
unresponsive and no logging will take place. This uses the same protocol as the Nextion editor and
only updates the changes of the TFT file. If HTTPS/SSL is enabled it will be about 1kB/sec.
```

.. warning::

>     If :ref:`uart-hardware_uarts` are not available than inconsistant results WILL occur. Lowering the baud to 9600 baud my help.

Reviewer 2: ```suggestion
    If :ref:`uart-hardware_uarts` are not available then inconsistent results WILL occur. Lowering the speed to 9600 baud may help.
```

> To host the TFT file you can use Home Assistant itself or any other web server. HTTPS while always recommened on any network will greatly reduce the upload speed.

Reviewer 2: ```suggestion
To host the TFT file you can use Home Assistant itself or any other web server. HTTPS, while always recommended on any network, will greatly reduce the upload speed.
```

Home Assistant
**************

> To host the TFT file from Home Assistant create a www directory if it doesnt exit in your config 

Reviewer 2: ```suggestion
To host the TFT file from Home Assistant, create a www directory if it doesn't exist in your config 
```

directory. You can create a subdirectory for those files as well.

For example if the file is located
under your configuration directory ``www/tft/default.tft`` the URL to access it will be
``http(s)://your_home_assistant_url:port/local/tft/default.tft``

NGINX
*****

`NGINX <https://www.nginx.com/>`__

> The below NGINX example configuration will server files out of /var/www/nextion directory.

Reviewer 2: ```suggestion
The below NGINX example configuration will serve files out of the /var/www/nextion directory.
```

.. code-block:: nginx

    server {
      listen 80;    
      access_log  /var/log/nginx/nextion_access.log;    
      error_log  /var/log/nginx/nextion_error.log;
      root /var/www/nextion;
    }

Components
----------
This library supports a few different components allowing communication back and forth from HA <-> MCU <-> Nextion.

.. note::

>     If the Nextion is sleeping or if the componet was set to be hidden it will not update its components even if updates are sent. After the Nextion wakes up all components will send their states to the Nextion to get around this.

Reviewer 2: ```suggestion
    If the Nextion is sleeping or if the component was set to be hidden, it will not update its components even if updates are sent.
    After the Nextion wakes up, all components will send their states to the Nextion to get around this.
```

> With the exception of the - :doc:`../binary_sensor/nextion` that has the ``page_id``/``component_id`` options configured, the example below illustrates:

Reviewer 2: ```suggestion
With the exception of the :doc:`../binary_sensor/nextion` that has the ``page_id``/``component_id`` options configured, the example below illustrates:
```
